File: backend/src/sor/models/project.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import StrEnum
from uuid import uuid4

from pydantic import BaseModel, Field

from .stage import StageResult


def _new_id() -> str:
    return uuid4().hex[:12]


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class ProjectState(StrEnum):
    DRAFT = "draft"
    IN_PROGRESS = "in_progress"
    COMPLETE = "complete"


class Project(BaseModel):
    id: str = Field(default_factory=_new_id)
    name: str
    research_question: str
    context: str = ""
    state: ProjectState = ProjectState.DRAFT
    current_stage: int = 1
    stage_results: list[StageResult] = Field(default_factory=list)
    created_at: str = Field(default_factory=_now_iso)
    updated_at: str = Field(default_factory=_now_iso)

File: ios/SocietyOfResearchers/Models/Project.swift
import Foundation

enum ProjectState: String, Codable, Sendable {
    case draft
    case inProgress = "in_progress"
    case complete
}

struct Project: Codable, Identifiable, Sendable {
    let id: String
    var name: String
    var researchQuestion: String
    var context: String
    var state: ProjectState
    var currentStage: Int
    var stageResults: [StageResult]
    var createdAt: String
    var updatedAt: String

    enum CodingKeys: String, CodingKey {
        case id, name, context, state
        case researchQuestion = "research_question"
        case currentStage = "current_stage"
        case stageResults = "stage_results"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct CreateProjectRequest: Encodable, Sendable {
    let name: String
    let researchQuestion: String
    var context: String = ""

    enum CodingKeys: String, CodingKey {
        case name, context
        case researchQuestion = "research_question"
    }
}

struct UpdateProjectRequest: Encodable, Sendable {
    var name: String?
    var researchQuestion: String?
    var context: String?

    enum CodingKeys: String, CodingKey {
        case name, context
        case researchQuestion = "research_question"
    }
}
